File: go-manychain/network/local_connection.go
package network

// LocalConnection implements IConnection
type LocalConnection struct {
	endpoint LocalEndpoint
	stream   LocalStream
}

func NewLocalConnection(endpoint LocalEndpoint) *LocalConnection {
	return &LocalConnection{
		endpoint: endpoint,
		stream:   *NewLocalStream(),
	}
}

// Send is an implementation for IConnection
func (c *LocalConnection) Send(payload RPC) {
	c.stream.Send(payload)
}

// Receive is an implementation for IConnection
func (c *LocalConnection) Receive(action func(RPC)) {
	c.stream.Receive(action)
}

// Close is an implementation for IConnection
func (c *LocalConnection) Close() {
	c.stream.Close()
}

func (c *LocalConnection) Name() string {
	return c.endpoint.Name()
}

File: rust-manychain/src/network/connections/local_connection.rs
use crate::network::connections::i_connection::IConnection;
use crate::network::endpoints::network_endpoint_local::LocalNetworkEndpoint;
use crate::network::stream::Stream;
use crate::network::streams::stream_local::LocalStream;

pub struct LocalConnection{
    endpoint: LocalNetworkEndpoint,
    stream: LocalStream
}

impl LocalConnection {
    pub fn new<'a>(endpoint: LocalNetworkEndpoint) ->  LocalConnection {
        LocalConnection{
            endpoint: endpoint.clone(),
            stream: LocalStream::new(endpoint.clone())
        }
    }
}

impl IConnection for LocalConnection {
    type T = LocalNetworkEndpoint;
    async fn send_message(&self, payload: Vec<u8>) {
        self.stream.send_message(payload).await;
    }
    async fn consume(&mut self) {
        self.stream.consume().await;
    }
}
